source/utils/adios_iotest/adios_remote_read.cpp

```diff
@@ -0,0 +1,672 @@
+                    var.SetSelection(adios2::Box<adios2::Dims>({startX}, {countX}));
+                    reader.Get<double>(var, data1D.data(), adios2::Mode::Sync);
+
+                    if (DEBUG)
+                        out = getOutputString(rank, output_line_length, out, data1D);
+                }
+            }
+
+            reader.EndStep();
+        }
+
+        reader.Close();
+        end = std::chrono::system_clock::now();
+        std::chrono::duration<double> elapsed_seconds = end - start;
+        out += "Duration :" + std::to_string(elapsed_seconds.count()) + "\n";
+        std::ofstream out_file;
+        out_file.open("output1D-" + std::to_string(rank) + ".log");
+        out_file << out;
+        out_file.close();
+    }
+    catch (std::invalid_argument &e)
+    {
+        if (rank == 0)
+        {
+            std::cout << "Invalid argument exception, STOPPING PROGRAM\n";
+            std::cout << e.what() << "\n";
+        }
+    }
+    catch (std::ios_base::failure &e)
+    {
+        if (rank == 0)
+        {
+            std::cout << "System exception, STOPPING PROGRAM\n";
+            std::cout << e.what() << "\n";
+        }
+    }
+    catch (std::exception &e)
+    {
+        if (rank == 0)
+        {
+            std::cout << "Exception, STOPPING PROGRAM\n";
+            std::cout << e.what() << "\n";
+        }
+    }
+}
+
+std::string &getOutputString(int rank, int output_line_length, std::string &out,
+                             std::vector<double> &data)
+{
+    {
+        out += "rank " + std::to_string(rank) + ":" + "\n";
+        int counter = 0;
+
+        for (auto v : data)
+        {
+            counter++;
+            out += std::to_string(v) + " ";
+            if (counter == output_line_length)
+            {
+                out += "\n";
+                counter = 0;
+            }
+        }
+    }
+    return out;
+}
+
+void read3D(int nproc, int rank, const std::string &filename, const int NSTEPS, adios2::IO &io,
+            std::vector<std::string> &variables_in, int direction, double ratio,
+            int output_line_length)
+{
+    unsigned int startX;
+    unsigned int startY;
+    unsigned int startZ;
+    unsigned int countX;
+    unsigned int countY;
+    unsigned int countZ;
+
+    try
+    {
+        std::chrono::time_point<std::chrono::system_clock> start_time;
+        std::chrono::time_point<std::chrono::system_clock> end_time;
+        start_time = std::chrono::system_clock::now();
+        // Get io settings from the config file or
+        // create one with default settings here
+
+        /*
+         * Define global array: type, name, global dimensions
+         * The local process' part (start, count) can be defined now or later
+         * before Write().
+         */
+
+        adios2::Engine reader = io.Open(filename, adios2::Mode::Read);
+
+        std::string out;
+        for (size_t step = 0; step < NSTEPS; step++)
```

Comment:
The number of steps should be an input parameter, and, if missing, all steps in the BP files should be read.
Unfortunately the C++ function which retrieves the number of steps (like adios2_steps) is missing. We can change it like that and quite the loop on NSTEPS counter.

Reply:
Well, it isn't missing, but that still doesn't mean you should use it.  Engine::Steps() can only return useful information for file engines because streaming engines don't know when the stream will end until the writer closes it.  (Generally I'd be surprised to find anything missing from the C++ interface that is present in others because most things are implemented in terms of C++.)

Reply:
If the number of steps written is lower than your input variable I think this code would crash so I would still use what is returned by BeginStep in the for and break is you reached the desired number of steps.